src/AggregateFunctions/AggregateFunctionGroupConcat.cpp

```diff
@@ -0,0 +1,297 @@
+
+                cur_data.is_empty = false;
+                cur_data.offsets.push_back(cur_data.size, arena);
+                cur_data.insertChar(rhs_data.data + rhs_data.getString(i), rhs_data.getSize(i), arena);
+                cur_data.count++;
+                cur_data.offsets.push_back(cur_data.size, arena);
+            }
+        }
+        else
+        {
+            if (!cur_data.is_empty)
+                cur_data.insertStr(delimiter, arena);
+
+            cur_data.is_empty = false;
+            cur_data.insertChar(rhs_data.data, rhs_data.size, arena);
+        }
+    }
+
+    void serialize(ConstAggregateDataPtr __restrict place, WriteBuffer & buf, std::optional<size_t> /* version */) const override
+    {
+        auto & cur_data = this->data(place);
+
+        writeVarUInt(cur_data.size, buf);
+        writeVarUInt(cur_data.allocated_size, buf);
+
+        writeBinary(cur_data.is_empty, buf);
+
+        buf.write(cur_data.data, cur_data.size);
+
+        if constexpr (has_limit)
+        {
+            writeVarUInt(cur_data.count, buf);
+            for (const auto & offset : cur_data.offsets)
+                writeVarUInt(offset, buf);
+        }
+    }
+
+    void deserialize(AggregateDataPtr __restrict place, ReadBuffer & buf, std::optional<size_t> /* version */, Arena * arena) const override
+    {
+        auto & cur_data = this->data(place);
+
+        readVarUInt(cur_data.size, buf);
+        readVarUInt(cur_data.allocated_size, buf);
+
+        readBinary(cur_data.is_empty, buf);
+
+        buf.readStrict(cur_data.data, cur_data.size);
+
+        if constexpr (has_limit)
+        {
+            readVarUInt(cur_data.count, buf);
+            cur_data.offsets.resize_exact(cur_data.count * 2, arena);
+            for (auto & offset : cur_data.offsets)
+                readVarUInt(offset, buf);
+        }
+    }
+
+    void insertResultInto(AggregateDataPtr __restrict place, IColumn & to, Arena *) const override
+    {
+        auto & cur_data = this->data(place);
+
+        if (cur_data.is_empty)
+        {
+            auto & column_string = assert_cast<ColumnString &>(to);
+            column_string.insert(NULL);
+            return;
+        }
+
+        auto & column_string = assert_cast<ColumnString &>(to);
+        column_string.insertData(cur_data.data, cur_data.size);
+    }
+
+    bool allocatesMemoryInArena() const override { return true; }
+};
+
+AggregateFunctionPtr createAggregateFunctionGroupConcat(
+    const std::string & name, const DataTypes & argument_types, const Array & parameters, const Settings *)
+{
+    assertUnary(name, argument_types);
+
+    bool has_limit = false;
+    UInt64 max_elems = 0;
+    String delimiter;
+
+    if (parameters.size() > 2)
+        throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH,
+            "Incorrect number of parameters for aggregate function {}, should be 0, 1 or 2", name);
+
+    if (!parameters.empty())
```

Comment:
We should also add check if parameters are empty, and return an exception with error code TOO_FEW_ARGUMENTS_FOR_FUNCTION

Reply:
Zero parameters it fine, in this case delimiter is empty string